flow/connectors/connelasticsearch/elasticsearch.go

```diff
@@ -78,101 +86,168 @@ func (esc *ElasticsearchConnector) Close() error {
+					}
+				}
 			}
 		}
 	}()
 
-	for qRecord := range stream.Records {
-		qRecordJsonMap := make(map[string]any)
+	var bulkIndexFatalError error
+	var bulkIndexErrors []error
+	var bulkIndexOnFailureMutex sync.Mutex
 
-		if upsertColIndex >= 0 {
-			docId = fmt.Sprintf("%v", qRecord[upsertColIndex].Value())
-		} else {
-			docId = uuid.New().String()
+	for record := range req.Records.GetRecords() {
+		var bodyBytes []byte
+		var err error
+		action := actionIndex
+
+		switch record.(type) {
+		case *model.InsertRecord[model.RecordItems], *model.UpdateRecord[model.RecordItems]:
+			bodyBytes, err = recordItemsProcessor(record.GetItems())
+			if err != nil {
+				esc.logger.Error("[es] failed to json.Marshal record", slog.Any("error", err))
+				return nil, fmt.Errorf("[es] failed to json.Marshal record: %w", err)
+			}
+		case *model.DeleteRecord[model.RecordItems]:
+			action = actionDelete
+			// no need to supply the document since we are deleting
+			bodyBytes = nil
 		}
-		for i, field := range schema.Fields {
-			switch r := qRecord[i].(type) {
-			// JSON is stored as a string, fix that
-			case qvalue.QValueJSON:
-				qRecordJsonMap[field.Name] = json.RawMessage(shared.
-					UnsafeFastStringToReadOnlyBytes(r.Val))
-			default:
-				qRecordJsonMap[field.Name] = r.Value()
+
+		bulkIndexer, ok := esBulkIndexerCache[record.GetDestinationTableName()]
+		if !ok {
+			bulkIndexer, err = esutil.NewBulkIndexer(esutil.BulkIndexerConfig{
+				Index:  record.GetDestinationTableName(),
+				Client: esc.client,
+				// can't really ascertain how many tables present to provide a reasonable value
+				NumWorkers:    1,
+				FlushInterval: 10 * time.Second,
+			})
+			if err != nil {
+				esc.logger.Error("[es] failed to initialize bulk indexer", slog.Any("error", err))
+				return nil, fmt.Errorf("[es] failed to initialize bulk indexer: %w", err)
 			}
+			esBulkIndexerCache[record.GetDestinationTableName()] = bulkIndexer
 		}
-		qRecordJsonBytes, err := json.Marshal(qRecordJsonMap)
+
+		tablePkey, err := model.RecToTablePKey(req.TableNameSchemaMapping, record)
 		if err != nil {
-			esc.logger.Error("[es] failed to json.Marshal record", slog.Any("error", err))
-			return 0, fmt.Errorf("[es] failed to json.Marshal record: %w", err)
+			esc.logger.Error("[es] failed to process record", slog.Any("error", err))
+			return nil, fmt.Errorf("[es] failed to process record: %w", err)
 		}
 
-		err = esBulkIndexer.Add(ctx, esutil.BulkIndexerItem{
-			Action:     "index",
-			DocumentID: docId,
-			Body:       bytes.NewReader(qRecordJsonBytes),
+		err = bulkIndexer.Add(ctx, esutil.BulkIndexerItem{
+			Action:     action,
+			DocumentID: tablePkey.String(),
+			Body:       bytes.NewReader(bodyBytes),
 
+			OnSuccess: func(_ context.Context, _ esutil.BulkIndexerItem, _ esutil.BulkIndexerResponseItem) {
+				recordCountsUpdateMutex.Lock()
+				defer recordCountsUpdateMutex.Unlock()
+				record.PopulateCountMap(tableNameRowsMapping)
```

Comment:
PopulateCountMap could be changed from `map[string]*struct{int,int,int}` to `map[string]struct{*atomicint,*atomicint,*atomicint}` & then no lock would be necessary by using atomics here

Reply:
can leave as is for now